### ISSUE TITLE
ACER - Examples on continuous action space

### ISSUE THREAD

**Issue by PKramek** [pfnet/pfrl#143]
Hello,

I am working on an RL project, where I want to use the ACER algorithm on continuous action space problems (Pybullet environments), but I have difficulties implementing it using Your framework. Would it be possible for You to add an example of how to use this algorithm on this class of problems?

**Comment by PKramek** [comment]
Specifically, I would like to know how to use models that do not share parameters across policy and value function

**Comment by muupan** [comment]
You are right, we don't have an example script for ACER with a continuous action space for now. It is nice to add one.

For now we only have a unit test for ACER with a continuous action space. Here is how the model is defined:
https://github.com/pfnet/pfrl/blob/44bf2e483f5a2f30be7fd062545de306247699a1/tests/agents_tests/test_acer.py#L388-L400
https://github.com/pfnet/pfrl/blob/44bf2e483f5a2f30be7fd062545de306247699a1/tests/agents_tests/test_acer.py#L409-L413

So `model`'s first `Linear` layer is shared. To implement a model with no parameters shared, I guess you can directly use `ACERContinuousActionHead` as a model, where `pi`, `v`, and `adv` do not share parameters.

**Comment by PKramek** [comment]
I get an error message when I try to use GaussianHeadWithFixedCovariance (as described in the original paper). Is that intentional?

`...pfrl/agents/acer.py", line 217, in compute_loss_with_kl_constraint
    assert param.requires_grad`

Here is my model definition for error reproduction

```
    dummy_env = gym.make(args.env)
    obs_size = dummy_env.observation_space.shape[0]
    action_size = dummy_env.action_space.shape[0]
    hidden_size = 256

    head = acer.ACERContinuousActionHead(
        pi=nn.Sequential(
            nn.Linear(hidden_size, action_size),
            pfrl.policies.GaussianHeadWithFixedCovariance(scale=0.3),
        ),
        v=nn.Sequential(
            nn.Linear(hidden_size, 1),
        ),
        adv=nn.Sequential(
            ConcatObsAndAction(),
            nn.Linear(hidden_size + action_size, 1),
        ),
    )

    model = nn.Sequential(
        nn.Linear(obs_size, hidden_size),
        nn.ReLU(),
        nn.Linear(hidden_size, hidden_size),
        nn.ReLU(),
        nn.Linear(hidden_size, hidden_size),
        nn.ReLU(),
        head,
    )

    model.apply(init_chainer_default)

    opt = pfrl.optimizers.SharedRMSpropEpsInsideSqrt(
        model.parameters(), lr=10e-5, eps=4e-3, alpha=0.99
    )

```

**Comment by muupan** [comment]
It seems like a bug. Can you try commenting out `assert param.requires_grad` and see if it works?

**Comment by PKramek** [comment]
I tried that and this causes the program to throw another exception: 

```
  File ".../venv/lib/python3.7/site-packages/pfrl/agents/acer.py", line 448, in compute_one_step_pi_loss
    delta=self.trust_region_delta,
  File ".../venv/lib/python3.7/site-packages/pfrl/agents/acer.py", line 222, in compute_loss_with_kl_constraint
    [original_loss], distrib_params, retain_graph=True
  File ".../venv/lib/python3.7/site-packages/torch/autograd/__init__.py", line 157, in grad
    inputs, allow_unused)
RuntimeError: One of the differentiated Tensors does not require grad
```

**Comment by PKramek** [comment]
Maybe Acer is trying to update its parameters before any data was propagated through its neural networks and it causes that exception.

**Comment by PKramek** [comment]
Thank you for a workaround idea, would it be possible for You to create a fix in the nearest future?

**Comment by PKramek** [comment]
And I don`t see learning rate used anywhere in `GaussianHeadWithStateIndependentCovariance` but from what I understand I could just pass a callable as `var_func`, which always returns desired variance value

**Comment by PKramek** [comment]
Could you point me out to some examples of how could I use separate PyTorch optimizers or rather just learning rates for my NNs in pfrl. In all of the provided examples, where there are multiple optimizers, they are passed to the training function as separate parameters. What If I wanted to set a different learning rate for my actor, and a different learning rate for my critic and at the same time set learning rate for just `GaussianHeadWithStateIndependentCovariance` layer in actor. Currently, I am using `pfrl.optimizers.SharedRMSpropEpsInsideSqrt` but it does not offer learning rate parametrization beyond setting just single value for all NNs

**Comment by muupan** [comment]
Here is PyTorch's official doc: https://pytorch.org/docs/stable/optim.html#per-parameter-options `pfrl.optimizers.SharedRMSpropEpsInsideSqrt` is a subclass of `torch.optim.Optimizer` so it is no different.

**Comment by muupan** [comment]
I hope #145 will resolve the issue.

**Comment by PKramek** [comment]
I will test it, I also would like to ask if Acer was tested on contiuous
action spaces. I ve tried training a few agents but did not see any
progress in achieved rewards, but it is very likely I ve made some mistakes
in hyperparametrization or learning process setup

czw., 27 maj 2021, 03:57 użytkownik Yasuhiro Fujita <
***@***.***> napisał:

**Comment by muupan** [comment]
It is tested on a toy env with continuous actions, but it is not verified that it can reproduce the performance on the continuous tasks in the paper.

Here is a sample script to train ACER on OpenAI Gym MuJoCo envs. It seems to work to some extent. It is not tuned much and I cannot guarantee that the hyperparameters etc. are same as those used in the paper.

https://github.com/muupan/pfrl/blob/acer-continous-example-tune/examples/mujoco/train_acer_mujoco.py

command:
```
python3 examples/mujoco/train_acer_mujoco.py --num-envs 8 --steps 3000000 --env HalfCheetah-v2
```
training log:
[scores.txt](https://github.com/pfnet/pfrl/files/6558621/scores.txt)

video with the final model:

https://user-images.githubusercontent.com/815970/119937435-b7ea9a00-bfc5-11eb-832c-4c7a0cb08d5b.mp4

**Comment by PKramek** [comment]
That really helped me, thank You very much for all the time You spend debugging code and helping me
